xdp/src/route.rs

```diff
@@ -0,0 +1,252 @@
+use {
+    crate::netlink::{
+        netlink_get_neighbors, netlink_get_routes, MacAddress, NeighborEntry, RouteEntry,
+    },
+    libc::{AF_INET, AF_INET6},
+    std::{
+        io,
+        net::{IpAddr, Ipv4Addr, Ipv6Addr},
+    },
+    thiserror::Error,
+};
+
+#[derive(Debug, Error)]
+pub enum RouteError {
+    #[error("no route found to destination {0}")]
+    NoRouteFound(IpAddr),
+
+    #[error("missing output interface in route")]
+    MissingOutputInterface,
+
+    #[error("could not resolve MAC address")]
+    MacResolutionError,
+}
+
+#[derive(Debug)]
+pub struct NextHop {
+    pub mac_addr: Option<MacAddress>,
+    pub ip_addr: IpAddr,
+    pub if_index: u32,
+}
+
+fn lookup_route(routes: &[RouteEntry], dest: IpAddr) -> Option<&RouteEntry> {
+    let mut best_match = None;
+
+    let family = match dest {
+        IpAddr::V4(_) => AF_INET as u8,
+        IpAddr::V6(_) => AF_INET6 as u8,
+    };
+
+    for route in routes.iter().filter(|r| r.family == family) {
+        match (dest, route.destination) {
+            // this is the default route
+            (_, None) => {
+                if best_match.is_none() {
+                    best_match = Some((route, 0));
+                }
+            }
+
+            (IpAddr::V4(dest_addr), Some(IpAddr::V4(route_addr))) => {
+                let prefix_len = route.dst_len;
+                if !is_ipv4_match(dest_addr, route_addr, prefix_len) {
+                    continue;
+                }
+
+                if best_match.is_none() || prefix_len > best_match.unwrap().1 {
+                    best_match = Some((route, prefix_len));
+                }
+            }
+
+            (IpAddr::V6(dest_addr), Some(IpAddr::V6(route_addr))) => {
+                let prefix_len = route.dst_len;
+                if !is_ipv6_match(dest_addr, route_addr, prefix_len) {
+                    continue;
+                }
+
+                if best_match.is_none() || prefix_len > best_match.unwrap().1 {
+                    best_match = Some((route, prefix_len));
+                }
+            }
+
+            // mixed address families - can't match
+            _ => continue,
+        }
+    }
+
+    best_match.map(|(route, _)| route)
+}
+
+fn is_ipv4_match(addr: Ipv4Addr, network: Ipv4Addr, prefix_len: u8) -> bool {
```

Comment:
these helpers can probably land in net-utils as they do not necessarily relate to XDP only.

Reply:
The reason these are here is that I want this crate to be standalone. The netlink code is copy pasted from aya. As soon as I'm happy with the API, I'll move everything back to aya and this crate will go away.

Reply:
Since you are maintaining aya anyway, seems reasonable to pull it in as a dep rather than having the same code stuck in 2 places.

Reply:
I think what he meant is that the netlink code from Aya was copy pasted, but also modified. These modifications are not yet mature enough to land in Aya, and he plans to move it back to Aya once fleshed out.

Reply:
Yes, but also everything else. I plan to move the proper XDP API back to aya (or a separate aya-af-xdp crate, haven't decided yet), so the whole crate will go away, or at least 99% of it.

Reply:
ok thats reasonable.